miner/miner.go

```diff
@@ -560,6 +565,37 @@ func (m *Miner) mineOne(ctx context.Context, base *MiningBase) (minedBlock *type
 		return nil, err
 	}
 
+	tEquivocateWait := build.Clock.Now()
+
+	// TODO: make param
+	m.niceSleep(time.Until(base.ComputeTime.Add(2 * time.Second)))
+	newBase, err := m.GetBestMiningCandidate(ctx)
+	if err != nil {
+		err = xerrors.Errorf("failed to refresh best mining candidate: %w", err)
+		return nil, err
+	}
+
+	// Only factor in equivocated blocks if doing so will not risk us missing a block
+	if newBase.TipSet.Height() == base.TipSet.Height() && newBase.TipSet.MinTicket().Equals(base.TipSet.MinTicket()) {
```

Comment:
Shouldn't this be checking if the tipsets are not equal?

Reply:
Wait, why are we looking at the min ticket at all?

Also, I realize there's a case we didn't consider: we switch to an entirely different tipset with a different set of parents. In that case, we can't look at the intersection... I'm not sure what we should do (ideally we'd be able to ask for the "best tipset with the current parents"...)

Reply:
Sorry, I should've explained my thinking here a lot more. My goal is to protect miner's chances of winning a block as much as possible, even if that allows for _some_ equivocation from other miners. Specifically, when only bother modifying our base tipset if all the work we've done so far (ticket scratching, etc.) is still valid. If the refreshed head invalidates any of our work, we just say "I'm choosing to ignore that", and submit the block anyway.

As a result, I think the only thing we need to look at here is the MinTicket -- if that's the same, all our work so far is still valid*, we can intersect our tipsets and submit our block on the intersection. If it's different, regardless of whether that's because we just have a new head, or because equivocation occurred in a way that caused the MinTicket to change, we submit our block as prepared.

This isn't necessarily the right thing to protect for, but I really don't want to increase the chance of missed blocks. Happy to hear arguments otherwise.

*all our work so far, with one exception that I hadn't though about: the messages we've selected might need to change. We can either just move message selection to after this check, or re-select messages in the unlikely case that we end up here. Thoughts?

Reply:
(But, yes, we should be short-circuiting if the tipsets are equal)

Reply:
That mostly makes sense but I have a few questions:

1. What if the new min ticket is actually better?
2. What if switching the base tipset doesn't matter?

I'm also concerned that this makes things easier to attack. E.g., if I think I'm winning, I can equivocate because I'll be the min ticket and nobody will switch away from that min ticket.